[TSan, SanitizerBinaryMetadata] Analyze the capture status for `alloca` rather than arbitrary `Addr`

Camsyn opened the pull request:
This PR is based on my last PR #132752 (the first commit of this PR), but addressing a different issue.

This commit addresses the limitation in `PointerMayBeCaptured` analysis when dealing with derived pointers (e.g. arr+1) as described in issue #132739.

The current implementation of `PointerMayBeCaptured` may miss captures of the underlying `alloca` when analyzing derived pointers, leading to some FNs in TSan, as follows:
```cpp
void *Thread(void *a) {
  ((int*)a)[1] = 43;
  return 0;
}

int main() {
  int Arr[2] = {41, 42};
  pthread_t t;
  pthread_create(&t, 0, Thread, &Arr[0]);
  // Missed instrumentation here due to the FN of PointerMayBeCaptured
  Arr[1] = 43;
  barrier_wait(&barrier);
  pthread_join(t, 0);
}
```
Refer to this [godbolt page](https://godbolt.org/z/n67GrxdcE) to get the compilation result of TSan.

Even when `PointerMayBeCaptured` working correctly, it should backtrack to the original `alloca` firstly during analysis, causing redundancy to the outer's `findAllocaForValue`.
```cpp
    const AllocaInst *AI = findAllocaForValue(Addr);
    // Instead of Addr, we should check whether its base pointer is captured.
    if (AI && !PointerMayBeCaptured(Addr, true)) ...
```

Key changes:
Directly analyze the capture status of the underlying `alloca` instead of derived pointers to ensure accurate capture detection
```cpp
    const AllocaInst *AI = findAllocaForValue(Addr);
    // Instead of Addr, we should check whether its base pointer is captured.
    if (AI && !PointerMayBeCaptured(AI, true)) ...
```

melver reviewed (changes requested):
I think your description is wrong:

> ```
>  const AllocaInst *AI = findAllocaForValue(Addr);
>  // Instead of Addr, we should check whether its base pointer is captured.
> if (AI && !PointerMayBeCaptured(Addr, AI)) ...
> ```

Shouldn't it say "if (AI && !PointerMayBeCaptured(AI, true))" ?

Camsyn commented:
> I think your description is wrong:
> 
> > ```
> >  const AllocaInst *AI = findAllocaForValue(Addr);
> >  // Instead of Addr, we should check whether its base pointer is captured.
> > if (AI && !PointerMayBeCaptured(Addr, AI)) ...
> > ```
> 
> Shouldn't it say "if (AI && !PointerMayBeCaptured(AI, true))" ?

Ohhh, it is a literal mistake in the PR description, but the relevant modification is correct.

Camsyn commented on llvm/test/Instrumentation/ThreadSanitizer/capture.ll:
```diff
@@ -88,4 +88,35 @@ entry:
 ; CHECK: __tsan_write
 ; CHECK: ret void
 
```
I'm sorry this PR has confused you 😿. 
In fact, these two tests in `capture.ll` are from my previous PR #132752, and have nothing to do with this PR. 

This PR comes with the commit of the previous PR because this PR needs to use its results (I am a novice in submitting a PR and don't know how to deal with this scenario).

The first commit should be eliminated after that PR is accepted.

melver replied:
I understood what's happening. It's an unfortunate consequence of the poor Github PR experience that LLVM project enforces :-/ -- normally I'd prefer a PR with multiple commits and not multiple PRs, but the fact we're forced to "Squash and Merge" makes this hard. Not your fault. Once the first PR is committed, you can rebase this one and the problem should be "fixed".

With regards to the 2nd commit in this PR, it'd still be good to have an IR test in addition to the e2e test you already added.

melver commented:
This looks good. But we should commit the other PR first, and rebase this one. I leave it to you when you want to submit the other PR or if you want to await other reviewers' feedback too.

Camsyn committed:
[TSan, SanitizerBinaryMetadata] Analyze the capture status for `alloca` rather than arbitrary `Addr`

This commit addresses the limitation in `PointerMayBeCaptured` analysis
when dealing with derived pointers (e.g. arr+1) as described in
issue #132739.

The current implementation may miss captures of the underlying alloca
when analyzing derived pointers. Even when working correctly,
backtracking to the original alloca during analysis causes redundancy
to the outer's `findAllocaForValue`.

Key changes:
Directly analyze the capture status of the underlying alloca instead
   of derived pointers to ensure accurate capture detection

This patch fixes some FNs of TSan, referring to the appending
testcases for more details.

Camsyn commented:
As the prerequisite PR #132752 has been merged, I rebased this PR to the main branch. I think it's okay to merge now.

Just an FYI - another NFC PR #132744 relevant to this PR has been merged, which highlights `PointerMayBeCaptured` usage scenarios in the comments (violated by TSan/SanitizerBinaryMetadata before this PR). 

Friendly ping :) @melver

nikic reviewed (approved):
LGTM
